syntax = "proto3";

package labs;

import "google/protobuf/timestamp.proto";

service LabService {
  // Labs Management
  rpc CreateLab (CreateLabRequest) returns (Lab);
  rpc GetLab (GetLabRequest) returns (Lab);
  rpc GetLabs (GetLabsRequest) returns (LabList);
  rpc UpdateLab (UpdateLabRequest) returns (Lab);
  rpc DeleteLab (DeleteLabRequest) returns (DeleteLabResponse);

  // Assets Management
  rpc UploadAsset (stream UploadAssetRequest) returns (Asset);
  rpc UpdateAsset (stream UpdateAssetRequest) returns (Asset);
  rpc DownloadAsset (DownloadAssetRequest) returns (stream DownloadAssetResponse);
  rpc DeleteAsset (DeleteAssetRequest) returns (DeleteAssetResponse);
  rpc ListAssets (ListAssetsRequest) returns (AssetList);
}

// Lab management
message Lab {
  int64 lab_id = 1;
  int64 owner_id = 2;
  string title = 3;
  google.protobuf.Timestamp created_at = 4;
  google.protobuf.Timestamp updated_at = 5;
  string abstract = 6;
  int64 submissions = 7;
  repeated int64 related_articles_ids = 8;
  repeated int32 tags_ids = 9;
}

message CreateLabRequest {
  int64 owner_id = 1;
  string title = 2;
  string abstract = 3;
  repeated int64 related_articles_ids = 4;
  repeated int32 tags_ids = 5;
}

message GetLabRequest {
  int64 lab_id = 1;
}

message GetLabsRequest {
  int32 page_number = 1;
  int32 page_size = 2;
}

message LabList {
  int64 total_count = 1;
  repeated Lab labs = 2;
}

message UpdateLabRequest {
  int64 lab_id = 1;
  string title = 2;
  string abstract = 3;
  repeated int64 related_articles_ids = 4;
  repeated int32 tags_ids = 5;
}

message DeleteLabRequest {
  int64 lab_id = 1;
}

message DeleteLabResponse {
  bool success = 1;
}

// Assets Management
message Asset {
  int64 asset_id = 1;
  int64 lab_id = 2;
  string filename = 3;
  int64 filesize = 4;
  google.protobuf.Timestamp upload_date = 5;
}

message UploadAssetMetadata {
  int64 lab_id = 1;
  string filename = 2;
  int64 filesize = 3;
}

message UploadAssetRequest {
  oneof data {
    UploadAssetMetadata metadata = 1;
    bytes chunk = 2;
  }
}

message UpdateAssetMetadata {
  int64 asset_id = 1;
  string filename = 2;
  int64 filesize = 3;
}

message UpdateAssetRequest {
  oneof data {
    UpdateAssetMetadata metadata = 1;
    bytes chunk = 2;
  }
}

message DownloadAssetRequest {
  int64 asset_id = 1;
}

message DownloadAssetResponse {
  oneof data {
    Asset asset = 1;
    bytes chunk = 2;
  }
}

message DeleteAssetRequest {
  int64 asset_id = 1;
}

message DeleteAssetResponse {
  bool success = 1;
}

message ListAssetsRequest {
  int64 lab_id = 1;
}

message AssetList {
  int64 total_count = 1;
  repeated Asset assets = 2;
}
